config/initializers/doorkeeper.rb

```diff
@@ -1,23 +1,46 @@
 # frozen_string_literal: true
 
+require "./lib/level_of_authentication"
+
 Doorkeeper.configure do
   # Change the ORM that doorkeeper will use (requires ORM extensions installed).
   # Check the list of supported ORMs here: https://github.com/doorkeeper-gem/doorkeeper#orms
   orm :active_record
 
   # This block will be called to check whether the resource owner is authenticated or not.
   resource_owner_authenticator do
+    if Rails.application.config.feature_flag_enforce_levels_of_authentication
+      level_scopes = request.params[:scope]
+                            .split(" ")
+                            .select { |scope| scope.starts_with?("level") }
+
+      required_auth_level = LevelOfAuthentication.select_highest_level(level_scopes)
+      current_auth_level = session[:level_of_authentication].to_s
+    else
+      required_auth_level = "level1"
+      current_auth_level = "level1"
+    end
+
     if current_user
+      unless LevelOfAuthentication.current_auth_greater_or_equal_to_required(current_auth_level, required_auth_level)
+        raise LevelOfAuthenticationTooLowError
+      end
+
       current_user
```

Comment:
Having a `current_auth_level` when the user may not be logged in feels a bit strange, how about something like:

```ruby
    required_level_of_authentication =
      LevelOfAuthentication.select_highest_level(
        request.params[:scope].split(" ").select { |scope| scope.starts_with?("level") },
      )

    if current_user
      current_level_of_authentication = session[:level_of_authentication]
      current_level_of_authentication_is_good_enough =
        !Rails.application.config.feature_flag_enforce_levels_of_authentication ||
        LevelOfAuthentication.current_auth_greater_or_equal_to_required(current_level_of_authentication, required_level_of_authentication)

      if current_level_of_authentication_is_good_enough
        current_user
      else
        raise LevelOfAuthenticationTooLowError
      end
```